fix of issue 196: exception safety (obj processors do not report validation errors when reloading some models):

goto40 opened the pull request:
* Problem was, that **with a global repository in the metamodel**, that **models with errors were cached** and reused (w/o beeing checked). Related docs: http://textx.github.io/textX/stable/scoping/#note-on-uniqueness-of-model-elements-global-repository
 * Solution: **delete all models touched while loading a model, when an error occurs** in object processors (in all repositories). - - > **strong exception safety guarantee**. 
 * _Alternative (not implemented)_: re-check cached models (Problem: what happens if the object processor modifies the model? Not good running object processors twice)
 * Minor: added (commented) printfs and commented some old printfs found in the codebase.

In principle this is a BIC. However, the previous behavior is a bug in my eyes... @igordejanovic: How to proceed?

<!-- Please don't remove/change code review checklist -->

## Code review checklist

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Commit messages are meaningful (see [this][commit messages] for details)
- [x] Tests have been included and/or updated (**used testcase sketched in #196**, thank you @danixeee )
- [x] Docstrings have been included and/or updated, as appropriate **N/A**
- [x] Standalone docs have been updated accordingly **N/A??**
- [ ] Changelog(s) has/have been updated, as needed (see `CHANGELOG.md`).


[commit messages]: https://chris.beams.io/posts/git-commit/

goto40 committed:
Analysis and possible fix:
 * Problem was with a global repository in the metamodel, that models with errors were cached and reused (w/o beeing checked).
 * Solution: delete all models touched while loading a model, when an error occurs in object processors (in all repositories).
 * Alternative (not implemented): re-check cached models (Problem: what happens if the object processor modifies the model? Not good running object processors twice)
 * Minor: added (commented) printfs and commented some old printfs found in the codebase.

danixeee commented:
There is still bug here. Please try this test:

```python
MODEL_WITH_IMPORT_ERROR = """
#include "data_structures_including_error.edata2"   // <--- types must be lowercase

#include "types.etype2"

algo A1 : Point -> City
algo A2 : City -> Population
connect A1 -> A2
"""

MODEL_WITH_TYPE_ERROR = """
#include "data_structures_including_error.edata2"
#include "types.etype2"

algo A1 : Point -> City
algo A2 : City -> Population
connect A1 -> A                                     // <--- Unknown object "A" of class "Algo"
"""


def test_flow():
    mmF = metamodel_for_language('flow-dsl-s')

    errors = []

    try:
        mmF.model_from_str(MODEL_WITH_IMPORT_ERROR)
    except TextXError as e:
        errors.append(e) # Caught: "types must be lowercase" (CORRECT)

    try:
        mmF.model_from_str(MODEL_WITH_TYPE_ERROR)
    except TextXError as e:
        errors.append(e) # Caught: "Unknown object "A" of class "Algo"" (CORRECT)

    try:
        mmF.model_from_str(MODEL_WITH_IMPORT_ERROR)
    except TextXError as e:
        errors.append(e) # Caught: "Unknown object "A" of class "Algo"" (WRONG)

    assert errors[0] == errors[2]
```

goto40 committed:
Added unittest: reproduced error observed by danixeee. Removing models seems not to work in all cases. Need analysis.

goto40 committed:
Fixed error. Not all cases catched (e.g. parser errors...) TBC next.

goto40 commented:
Thank you for the extended example. I found the error. I need more time to analyse other (similar) cases. I will check model cleanup on errors...

Your case seems to work now (**reason/explanation of your error/observation**: It was not a obj_processor error, but a scope provider error....).

goto40 committed:
flake8

goto40 committed:
removed printf

goto40 committed:
cleaned up cached models when a syntax error occurs.

goto40 committed:
flake8

goto40 commented:
Fixed behavior: We now clean up when
 * errors occur during parsing of included models,
 * during reference resolution, and
 * during object processor execution (as before)

Added tests.

goto40 commented:
@danixeee: Sorry for all these inconveniences. But I did not have the objective to continue to work with models or meta-models after an exception occurred... I fear you will get some other effects or bugs with all that logic (especially with cached models, when updating some files...).

goto40 commented:
Note: we now should have the **strong exception safety guarantee** for the activity 'load model'. See https://en.m.wikipedia.org/wiki/Exception_safety

danixeee commented:
@goto40 Just tried it now again and it works fine, thanks again for quick fix!

> Sorry for all these inconveniences. But I did not have the objective to continue to work with models or meta-models after an exception occurred...

One option was to reload meta-model every time before parsing, and that would happen on every key-press, so I think reusing meta-model is much better.

goto40 commented:
@danixeee 
> One option was to reload meta-model every time before parsing, and that would happen on every key-press, but I think reusing meta-model is much better.

I see a whole bunch effects which could happen when updating a model. Compare #199. 

For a first implementation it would be nice to reload the meta model every time you enter/leave the current editor widget (on_focus_changed or similar?). This forces to reload everything...

goto40 commented:
Close in favor of #200
